openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,201 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/kubernetes-incubator/external-storage/openstack/standalone-cinder/pkg/volumeservice"
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// ProvisionerName is the unique name of this provisioner
+	ProvisionerName = "openstack.org/standalone-cinder"
+
+	// ProvisionerIDAnn is an annotation to identify a particular instance of this provisioner
+	ProvisionerIDAnn = "standaloneCinderProvisionerIdentity"
+
+	// CinderVolumeID is an annotation to store the ID of the associated cinder volume
+	CinderVolumeID = "cinderVolumeId"
+)
+
+type cinderProvisioner struct {
+	// Openstack cinder client
+	VolumeService *gophercloud.ServiceClient
+
+	// Kubernetes Client. Use to create secret
+	Client kubernetes.Interface
+	// Identity of this cinderProvisioner, generated. Used to identify "this"
+	// provisioner's PVs.
+	Identity string
+
+	vsb volumeServiceBroker
+	mb  mapperBroker
+}
+
+// NewCinderProvisioner returns a Provisioner that creates volumes using a
+// standalone cinder instance and produces PersistentVolumes that use native
+// kubernetes PersistentVolumeSources.
+func NewCinderProvisioner(client kubernetes.Interface, id, configFilePath string) (controller.Provisioner, error) {
+	volumeService, err := volumeservice.GetVolumeService(configFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get volume service: %v", err)
+	}
+
+	return &cinderProvisioner{
+		VolumeService: volumeService,
+		Client:        client,
+		Identity:      id,
+		vsb:           &gophercloudBroker{},
+		mb:            &volumeMapperBroker{},
+	}, nil
+}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	var (
+		connection volumeservice.VolumeConnection
+		mapper     volumeMapper
+		pv         *v1.PersistentVolume
+		cleanupErr error
+	)
+
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	volumeID, err := p.vsb.CreateCinderVolume(p.VolumeService, options)
+	if err != nil {
+		glog.Errorf("Failed to create volume")
+		goto ERROR
+	}
+
+	err = p.vsb.WaitForAvailableCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Volume %s did not become available", volumeID)
+		goto ERROR_DELETE
```

Comment:
This comes from my days as a Linux kernel programmer :smile:  I think goto (when used properly) can produce very readable error cleanup code.  Since I am curious, would you implement a cleanup function like this:
```go
func cleanup(unreserve bool, disconnect bool, delete bool) {
    if unreserve {
        // unreserve volume
    }
    if disconnect {
        // disconnect volume
    }
    if delete {
        // delete volume
    }
}
```

Reply:
sgtm